presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CombineApproxPercentileFunctions.java

```diff
@@ -0,0 +1,278 @@
+    }
+
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(CombineApproxPercentileFunctions::hasMultipleApproxPercentile);
+
+    private static boolean hasMultipleApproxPercentile(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations().values().stream()
+                .filter(agg -> agg.getCall().getDisplayName().equals(APPROX_PERCENTILE)).count() > 1;
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isCombineApproxPercentileEnabled(session);
+    }
+
+    // Get the position of the percentile argument in arguments
+    private static int getPercentilePosition(FunctionHandle functionHandle)
+    {
+        // approx_percentile(x, percentage) -> arguments.size() == 2
+        // approx_percentile(x, percentage, accuracy) -> arguments.size() == 3 && arguments.get(1) is Double
+        // approx_percentile(x, w, percentage) -> arguments.size() == 3 && arguments.get(1) is BigInt
+        // approx_percentile(x, w, percentage, accuracy) -> arguments.size() == 4
+        if (functionHandle instanceof BuiltInFunctionHandle) {
+            List<TypeSignature> argumentTypes = ((BuiltInFunctionHandle) functionHandle).getSignature().getArgumentTypes();
+            if (argumentTypes.size() == 2
+                    || (argumentTypes.size() == 3 && argumentTypes.get(1).getBase().equals(StandardTypes.DOUBLE))) {
+                return 1;
+            }
+            else if (argumentTypes.size() == 4
+                    || (argumentTypes.size() == 3 && argumentTypes.get(1).getBase().equals(StandardTypes.BIGINT))) {
+                return 2;
+            }
+        }
+
+        checkState(false);
+        return -1;
+    }
+
+    private static List<RowExpression> addPercentileArgument(List<RowExpression> argumentsWithoutPercentile, RowExpression percentileArgument, int percentilePosition)
+    {
+        List<RowExpression> newAggCallArguments = new ArrayList<>();
+        newAggCallArguments.addAll(argumentsWithoutPercentile);
+        newAggCallArguments.add(percentilePosition, percentileArgument);
+        return newAggCallArguments;
+    }
+
+    // Aggregations which are the same after removing percentile arguments are candidates to be combined
+    private static AggregationNode.Aggregation aggregationWithoutPercentile(AggregationNode.Aggregation aggregation, int percentilePosition)
+    {
+        List<RowExpression> arguments = aggregation.getCall().getArguments();
+        ImmutableList.Builder<RowExpression> argumentWithoutPercentile = ImmutableList.builder();
+        for (int i = 0; i < arguments.size(); ++i) {
+            if (i != percentilePosition) {
+                argumentWithoutPercentile.add(arguments.get(i));
+            }
+        }
```

Comment:
Even if we have the same function handle, we still need to check individual arguments, for example accuracy, weight as they need to be the same too.

Reply:
> Even if we have the same function handle, we still need to check individual arguments, for example accuracy, weight as they need to be the same too. I will update the code based on our discussion offline.

Yes, my idea is to streamline the logic to first find candidates to merge and then apply the actual merge algo where you need additional checks so it will be more readable.

Reply:
Updated the code. 

I think the part which causes confusion is the code which removes percentile arguments from aggregations for comparison. Hence in this part, I refactor the code to directly compare the different fields of the aggregations instead of building dummy aggregations to make the code more readable.

In this version, when I tried to implement it, I found that after partition by column and function handle, I still need to check most part of the aggregation, considering that I refactor the code to remove the confusing part, I just directly do the comparison and populate the candidate aggregations in one loop. How do you think? Thanks!